### PR TITLE
recompile when circum src files changed

### PR DESCRIPTION
**Pull request by winedia**

### CODE REVIEW

File: src/witness_generator.ts [hunk 1]
```diff
@@ -111,6 +112,72 @@ async function compileNativeBinary({
...
+}
+
+function loadOldSrcHashes(src) {
+  const hashesFile = path.join(src, '..', 'srcs.sum');
+  try {
+    const jsonStr = fs.readFileSync(hashesFile, 'utf8');
+    const hashes = new Map(JSON.parse(jsonStr));
+    return hashes;
+  } catch (err) {
+    return null;
```
Comment (lispc) [inline src/witness_generator.ts:136]: console log err then return null

File: src/witness_generator.ts [hunk 2]
```diff
@@ -111,6 +112,72 @@ async function compileNativeBinary({
...
+  const jsonStr = JSON.stringify(Array.from(hashes.entries()));
+  const hashesFile = path.join(src, '..', 'srcs.sum');
+  fs.writeFileSync(hashesFile, jsonStr, 'utf8');
+}
+
+function isEqual(srcHashes, oldSrcHashes) {
+  if (srcHashes.length !== oldSrcHashes.length) {
+    return false;
+  }
+  for (const src of Array.from(srcHashes.keys())) {
```
Comment (lispc) [inline src/witness_generator.ts:150]: i suggest display **all** changed files so user can know why recompile happens

File: src/witness_generator.ts [hunk 3]
```diff
@@ -111,6 +112,72 @@ async function compileNativeBinary({
...
+  for (const entry of Array.from(contents.entries())) {
+    const checksum = crypto.createHash('md5').update(entry[1], 'utf8').digest('hex');
+    hashes.set(entry[0], checksum);
+  }
+  return hashes;
+}
+
+function loadOldSrcHashes(src) {
+  const hashesFile = path.join(src, '..', 'srcs.sum');
+  try {
```
Comment (lispc) [inline src/witness_generator.ts:128]: When the file does not exist return null directly. If some other unexpected error happens console log error and return null

File: src/witness_generator.ts [hunk 4]
```diff
@@ -111,6 +112,72 @@ async function compileNativeBinary({
...
+  }
+}
+
+function writeSrcHashes(hashes, src) {
+  const jsonStr = JSON.stringify(Array.from(hashes.entries()));
+  const hashesFile = path.join(src, '..', 'srcs.sum');
+  fs.writeFileSync(hashesFile, jsonStr, 'utf8');
+}
+
+function isEqual(srcHashes, oldSrcHashes) {
```
Comment (lispc) [inline src/witness_generator.ts]: Change function name maybe isSrcHashesEqual？

**Comment by lispc** [comment]
Besides add one or two lines of comments to explain these functions what is their purpose

**Review by 0xmountaintop — APPROVED** [review]

**Pull request merged**